os/fs/smartfs/smartfs_utils.c

```diff
@@ -1716,139 +1962,270 @@ int smartfs_countdirentries(struct smartfs_mountpt_s *fs, struct smartfs_entry_s
 		ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
 		if (ret < 0) {
-			fdbg("Error reading sector %d header, ret : %d\n", readwrite.logsector, ret);
-			return ret;
+			fdbg("Unable to read sector %d, error = %d\n", readwrite.logsector, ret);
+			break;
+		}
+		nextsector = SMARTFS_NEXTSECTOR(chainheader);
+		ret = FS_IOCTL(fs, BIOC_FREESECT, (unsigned long)readwrite.logsector);
+		if (ret < 0) {
+			fdbg("Unable to free sector %d to shrink file, error = %d\n", readwrite.logsector, ret);
+			break;
 		}
+	}
 
-		/* Get the next chained sector */
+	return ret;
+}
 
-		sector = SMARTFS_NEXTSECTOR(header);
+/****************************************************************************
+ * Name: smartfs_extendfile
+ *
+ * Description:
+ *   Zero-extend the length of a regular file to 'length'.
+ *
+ ****************************************************************************/
 
-		/* If this is the 1st sector of the file, then just overwrite
-		 * the sector data with the erased state value.  The underlying
-		 * SMART block driver will detect this and release the old
-		 * sector and create a new one with the new (blank) data.
-		 */
+int smartfs_extendfile(FAR struct smartfs_mountpt_s *fs, FAR struct smartfs_ofile_s *sf, off_t length)
+{
+	int ret;
+	ssize_t writelen;
+	uint16_t data_len;
+	char *buf = NULL;
 
-		if (nextsector == entry->firstsector) {
-#ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
+	if (length == 0) {
+		return OK;
+	}
 
-			/* When we have a sector buffer in use, simply skip the first sector */
+	smartfs_seek_internal(fs, sf, 0, SEEK_END);
+	length -= sf->entry.datalen;
+	data_len = fs->fs_llformat.availbytes - sizeof(struct smartfs_chain_header_s);
+	buf = (char *)kmm_malloc(data_len);
+	if (!buf) {
+		fdbg("Error allocating space for buffer\n");
+		return -ENOMEM;
+	}
+	memset(buf, '\0', data_len);
 
-			nextsector = sector;
-			continue;
+	while (length > 0) {
+		writelen = data_len;
+		if (writelen > length) {
+			writelen = length;
+		}
+		if ((fs->fs_llformat.availbytes != sf->curroffset) && (writelen > (fs->fs_llformat.availbytes - sf->curroffset))) {
+			writelen = (fs->fs_llformat.availbytes - sf->curroffset);
+		}
 
-#else
+		ret = smartfs_append_data(fs, sf, buf, writelen);
+		if (ret != writelen) {
+			fdbg("Unable to append data to file, ret : %d, write length : %d\n", ret, writelen);
+			ret = -EIO;
+			goto error_with_buf;
+		}
+		length -= ret;
+	}
+	ret = OK;
+
+error_with_buf:
+	free(buf);
+	return ret;
+}
+
+/****************************************************************************
+ * Name: smartfs_append_data
+ *
+ * Description:
+ *   Append data to the end of file, consisting of all NULL,('\0') characters
+ *
+ ****************************************************************************/
+
+ssize_t smartfs_append_data(FAR struct smartfs_mountpt_s *fs, FAR struct smartfs_ofile_s *sf, const char *buffer, size_t buflen)
```

Comment:
I will add verbose log here and smartfs_write as a below to help debug, so could you please modify parameter as I sent you?
```
		fvdbg("datalen : %d sector : %d offset : %d filepos : %d bytewrite : %d buflen : %d\n", sf->entry.datalen, sf->currsector, \
		sf->curroffset, sf->filepos, byteswritten, buflen);
```

Reply:
Added